fedimint-client/src/transaction/builder.rs

```diff
@@ -475,44 +481,36 @@ impl TransactionBuilder {
         let input_states = self
             .inputs
             .into_iter()
+            .filter(|bundle| !bundle.is_empty())
             .enumerate()
             .flat_map(|(bundle_idx, bundle)| {
-                let input_idxs = find_range_of_matching_items(&input_idx_to_bundle_idx, bundle_idx);
-                bundle.sms.into_iter().flat_map(move |sm| {
-                    if let Some(input_idxs) = input_idxs.as_ref() {
-                        (sm.state_machines)(OutPointRange::new(
-                            txid,
-                            IdxRange::from_inclusive(input_idxs.clone()).expect("can't overflow"),
-                        ))
-                    } else {
-                        // TODO: In the future `InputBundle` and `OutputBundle` should make empty
-                        // bundles impossible, so this should not be possible
-                        vec![]
-                    }
+                let input_idxs = find_range_of_matching_items(&input_idx_to_bundle_idx, bundle_idx)
+                    .expect("Non empty bundles must always have a match");
+                bundle.sm_gens.into_iter().flat_map(move |sm| {
+                    (sm.state_machines)(OutPointRange::new(
+                        txid,
+                        IdxRange::from_inclusive(input_idxs.clone()).expect("can't overflow"),
+                    ))
                 })
             });
 
-        let output_states =
-            self.outputs
-                .into_iter()
-                .enumerate()
-                .flat_map(|(bundle_idx, bundle)| {
-                    let output_idxs =
-                        find_range_of_matching_items(&output_idx_to_bundle_idx, bundle_idx);
-                    bundle.sms.into_iter().flat_map(move |sm| {
-                        if let Some(output_idxs) = output_idxs.as_ref() {
-                            (sm.state_machines)(OutPointRange::new(
-                                txid,
-                                IdxRange::from_inclusive(output_idxs.clone())
-                                    .expect("can't possibly overflow"),
-                            ))
-                        } else {
-                            // TODO: In the future `InputBundle` and `OutputBundle` should make
-                            // empty bundles impossible, so this should not be possible
-                            vec![]
-                        }
-                    })
-                });
+        let output_states = self
+            .outputs
+            .into_iter()
+            .filter(|bundle| !bundle.is_empty())
+            .enumerate()
+            .flat_map(|(bundle_idx, bundle)| {
+                let output_idxs =
+                    find_range_of_matching_items(&output_idx_to_bundle_idx, bundle_idx)
+                        .expect("Non empty bundles must always have a match");
```

Comment:
Enumerate after filter feels fragile, wherever that idx is used has to have the same filtering in place.

Reply:
@elsirion It looks fragile because it's actually a bug. The bundles are numbered without any filtering. :facepalm: It would lead to a panic if we actually used some empty bundles. Good catch.

Reply:
Added a test and fix.